openmp/libomptarget/plugins-nextgen/cuda/src/rtl.cpp

```diff
@@ -486,6 +494,16 @@ struct CUDADeviceTy : public GenericDeviceTy {
       Res = cuMemAllocManaged(&DevicePtr, Size, CU_MEM_ATTACH_GLOBAL);
       MemAlloc = (void *)DevicePtr;
       break;
+    case TARGET_ALLOC_DEVICE_NON_BLOCKING: {
+      CUstream Stream;
+      if (Res = cuStreamCreate(&Stream, CU_STREAM_NON_BLOCKING))
```

Comment:
You mentioned that using the `getStream(AsyncInfoTy ... )` method seems to lock up the system. I wonder why that is, there shouldn't be much of a difference? At least from what it is supposed to do.

Reply:
So, there is no AsyncInfoTy object here, but I am also confused why we can't simply call getStream and later return it again. Stream creation and destruction is not free, hence our resource pools.

Reply:
I'm not 100% sure, but I think it's because the resource pools rarely ever delete the created streams. This lead to some problems. I tried making a single stream that was shared but this lead to weird errors as well. Hence, this being the easiest way to get it to stop deadlocking in the short term.

Reply:
> So, there is no AsyncInfoTy object here, but I am also confused why we can't simply call getStream and later return it again. Stream creation and destruction is not free, hence our resource pools.

I mean, `getStream` requires an `AsyncInfoWrapperTy` reference (at least the one I found), no? Obviously nothing would stop you from having a local object that you can use to synchronize once your `malloc` is done. Or am I missing something fundamental here?

Reply:
Okay, so I think one problem is that the logic states that a waiting thread can check the RPC server. However, this logic isn't really prepared to handle when a thread handling the RPC server already comes in and waits. So we end up with some kind of infinite recursion. I'll need to think of some logic to prevent this.

Reply:
Further, if I use a single stream I get a single successful allocation from it, then the rest return `nullptr` for unknown reasons.